man/pkginfo.8.in
.TH pkginfo 8 "" "cards #VERSION#" ""
.SH NAME
pkginfo \- display software package information
.SH SYNOPSIS
\fBpkginfo [options]\fP
.SH DESCRIPTION
\fBpkginfo\fP is a \fIpackage management\fP utility, which displays
information about software packages that are installed on the system
or that reside in a particular directory.
.SH OPTIONS
.TP
.B "\-i, \-\-installed"
List installed packages and their version.
.TP
.B "\-d, \-\-details"
List all available details about the package.
.TP
.B "\-L \-\-libraries"
List all the runtime libraries owned by the specified <package>
.TP
.B "\-l, \-\-list <package|file>"
List files owned by the specified <package> or contained in <file>.
.TP
.B "\-o, \-\-owner <pattern>"
List owner(s) of file(s) matching <pattern>.
.TP
.B "\-f, \-\-footprint <file>"
Print footprint for <file>. This feature is mainly used by pkgmk(8)
for creating and comparing footprints.
.TP
.B "\-a, \-\-archive <file>"
Print name, version, Release, BuildDate and dependencies of the <file>
.TP
.B "\-b, \-\-buildtime <package>"
Print on a single line the name and the build time of the package.
.TP
.B "\-R, \-\-runtimedep <package>"
Print on a single line all the runtime dependencies.
.TP
.B "\-\-runtimedepfiles <path>"
Print on a single line all the runtime dependencies for the files found in the <path>
.TP
.B "\-r, \-\-root <path>"
Specify alternative installation root (default is "/"). This
should be used if you want to display information about a package
that is installed on a temporary mounted partition, which is "owned"
by another system. By using this option you specify which package
database to use.
.TP
.B "\-v, \-\-version"
Print version and exit.
.TP
.B "\-h, \-\-help"
Print help and exit.
.SH SEE ALSO
pkgadd(8), pkgrm(8), pkgmk(8), rejmerge(8)
.SH COPYRIGHT
This program is part of the *p_cards package* Copyright (c) 2016, **peter1000** <https://github.com/peter1000>  
which is licensed under the GNU General Public License.
*p_cards* includes code derived from other open source projects, which have their own licenses:
Read the LICENSE.md file for the detailed license information.


CHANGELOG.md
# v1.99.9 (xxxx-xx-xx)

## Summary

* Renamed to `p_cards` peter1000 development repo
* Updated: CHANGELOG.md, LICENSE.md, README.md, TODO.md
* Adds: initial documentation structure for [MkDocs](https://github.com/mkdocs/mkdocs)
* Removes french manpages
* Removes sub-package *fltk*
* Replaces `pkgmk.in` with `pkgmk2.in`
* Removes individual license headers in favour of the overall package license.
* Adds *Christophe Devine*, *Junio C Hamano* and *Logan Ingalls* to the Copyright holder in LICENSE.md

# History

* On 07. March 2016 **peter1000** <https://github.com/peter1000/> renamed 
    `cards commit 89580529259f5e912a2efc936bebf90e3f698537` to **p_cards**.

* [cards-1.99.8](https://github.com/NuTyX/cards/commit/992a2ca11bf771e0809ebe9d0987bdc1d47a8e82)    (on Mar 03, 2016)
* [cards-1.99.0](https://github.com/NuTyX/cards/commit/d9d4ae224175ab8514b1951a0078cc0fb75d60fc)    (on Jan 14, 2016)
* [cards-1.96.0](https://github.com/NuTyX/cards/commit/ae6c74e2f2f463bea0699ab2bb43f5daa3df15f2)    (on Jan 03, 2016)
* [cards-1.95.0](https://github.com/NuTyX/cards/commit/9f37977d754bd1dd88346f28e50ea629a69034d8)    (on Jan 02, 2016)
* [cards-1.4.0](https://github.com/NuTyX/cards/commit/f7bd5a8d341eb63aa0a772a5d8e3c201d2e97b15)     (on Dec 25, 2015)
* [cards-1.3.0](https://github.com/NuTyX/cards/commit/a695725e37beb64bda56ca000c5ee86d1f34541e)     (on Aug 25, 2015)
* [cards-1.2.1](https://github.com/NuTyX/cards/commit/14b580fc4165f256e952262914d9cb7e2cba6f74)     (on Jul 23, 2015)
* [cards-1.1](https://github.com/NuTyX/cards/commit/f3eb1243dabec33f7854dd6330a53438d62c8440)       (on Jun 06, 2015)
* [cards-1.0](https://github.com/NuTyX/cards/commit/49452594a9fc8c312fdc4c7203d76761ec5eab9f)       (on May 14, 2015) 


src/pkginfo.cpp
#include "pkginfo.h"
#include <iterator>
#include <vector>
#include <iomanip>
#include <sys/types.h>

#include <stdio.h>
#include <unistd.h>
#include <sstream>

Pkginfo::Pkginfo(const string& commandName)
	: Pkgdbh(commandName),
	m_runtimedependencies_mode(0),
	m_footprint_mode(0),
	m_archiveinfo(0),
	m_installed_mode(0),
	m_list_mode(0),
	m_owner_mode(0),
	m_details_mode(0),
	m_libraries_mode(0),
	m_runtime_mode(0),
	m_epoc(0)
{
}
Pkginfo::Pkginfo()
	:Pkgdbh("pkginfo"),
	m_runtimedependencies_mode(0),
	m_footprint_mode(0),
	m_archiveinfo(0),
	m_installed_mode(0),
	m_list_mode(0),
	m_owner_mode(0),
	m_details_mode(0),
	m_libraries_mode(0),
	m_runtime_mode(0),
	m_epoc(0)
{
}
void Pkginfo::run(int argc, char** argv)
{
	parseArguments(argc,argv);
	run();
}
void Pkginfo::parseArguments(int argc, char** argv)
{
	for (int i = 1; i < argc; ++i) {
		string option(argv[i]);
		if (option == "-r" || option == "--root") {
			assertArgument(argv, argc, i);
			m_root = argv[i + 1];
			i++;
		} else if (option == "-i" || option == "--installed") {
			m_installed_mode += 1;
		} else if (option == "-d" || option == "--details") {
			assertArgument(argv, argc, i);
			m_details_mode +=1;
			m_arg = argv[i + 1];
			i++;
		} else if (option == "-l" || option == "--list") {
			assertArgument(argv, argc, i);
			m_list_mode += 1;
			m_arg = argv[i + 1];
			i++;
		} else if (option == "-R" || option == "--runtimedep") {
			assertArgument(argv, argc, i);
			m_runtime_mode += 1;
			m_arg = argv[i + 1];
			i++;
		} else if (option == "-o" || option == "--owner") {
			assertArgument(argv, argc, i);
			m_owner_mode += 1;
			m_arg = argv[i + 1];
			i++;
		} else if (option == "-a" || option == "--archive") {
			assertArgument(argv, argc, i);
			m_archiveinfo += 1;
			m_arg = argv[i + 1];
			i++;
		} else if (option == "-f" || option == "--footprint") {
			assertArgument(argv, argc, i);
			m_footprint_mode += 1;
			m_arg = argv[i + 1];
			i++;
		} else if (option == "-L" || option == "--libraries") {
			assertArgument(argv, argc, i);
			m_libraries_mode +=1;
			m_arg =  argv[i + 1];
			i++;
		} else if (option == "--runtimedepfiles") {
			assertArgument(argv, argc, i);
			m_runtimedependencies_mode +=1;
			m_arg = argv[i + 1];
			i++;
		} else if (option == "-b" || option == "--buildtime") {
			assertArgument(argv, argc, i);
			m_epoc +=1;
			m_arg = argv[i + 1];
			i++; 
		} else {
			m_actualError = INVALID_OPTION;
			treatErrors(option);
		}
		if (m_root.empty())
			m_root="/";
		else
			m_root=m_root+"/";

	}

	m_packageArchiveName = m_arg;

	if (m_runtimedependencies_mode + m_footprint_mode + m_details_mode +
	m_installed_mode + m_list_mode + m_owner_mode + m_epoc + m_archiveinfo +
	m_footprint_mode + m_libraries_mode + m_runtime_mode == 0)
	{
		m_actualError = OPTION_MISSING;
		treatErrors(m_arg);
	}
	if (m_runtimedependencies_mode + m_footprint_mode + m_installed_mode + m_archiveinfo + m_list_mode + m_owner_mode > 1)
	{
		m_actualError = TOO_MANY_OPTIONS;
		treatErrors(m_arg);
	}
}
void Pkginfo::run()
{
	if (m_archiveinfo) {
		pair<string, pkginfo_t> packageArchive = openArchivePackage(m_packageArchiveName) ;
		cout	<< packageArchive.first << " Description    : " << packageArchive.second.description << endl
			<< packageArchive.first << " URL            : " << packageArchive.second.url << endl
			<< packageArchive.first << " Maintainer(s)  : " << packageArchive.second.maintainer << endl
			<< packageArchive.first << " Packager(s)    : " << packageArchive.second.packager << endl
			<< packageArchive.first << " Version        : " << packageArchive.second.version << endl
			<< packageArchive.first << " Release        : " << packageArchive.second.release << endl
			<< packageArchive.first << " Architecture   : " << packageArchive.second.arch  << endl
			<< packageArchive.first << " Build date     : " << packageArchive.second.build << endl;
		if (packageArchive.second.dependencies.size() > 0 ) {
			cout << packageArchive.first << " Dependencies   : ";
			for ( auto i : packageArchive.second.dependencies) cout << i.first << i.second << " ";
			cout << endl;
		}
	}
	// Make footprint
	if (m_footprint_mode) {
		getFootprintPackage(m_arg);
	} else {
		// Modes that require the database to be opened
		Db_lock lock(m_root, false);
		getListOfPackageNames(m_root);
		if (m_installed_mode) {
			// List installed packages
			buildDatabaseWithNameVersion();
			for (auto i : m_listOfInstPackages) {
				cout << i.first << " " << i.second.version << "-" << i.second.release << endl;
			}
		} else if (m_list_mode) {
			// List package or file contents
			buildDatabaseWithDetailsInfos(false);
			if (checkPackageNameExist(m_arg)) {
				copy(m_listOfInstPackages[m_arg].files.begin(), m_listOfInstPackages[m_arg].files.end(), ostream_iterator<string>(cout, "\n"));
			} else if (checkFileExist(m_arg)) {
				pair<string, pkginfo_t> package = openArchivePackage(m_arg);
				copy(package.second.files.begin(), package.second.files.end(), ostream_iterator<string>(cout, "\n"));
			} else {
				m_actualError = NOT_INSTALL_PACKAGE_NEITHER_PACKAGE_FILE;
				treatErrors(m_arg);
			}
		} else if (m_runtimedependencies_mode) {
			/* 	Get runtimedependencies of the file found in the directory path
				get the list of installed package silently */
			buildDatabaseWithDetailsInfos(true);
			regex_t r;
			int Result;
			regcomp(&r, ".", REG_EXTENDED | REG_NOSUB);
			set<string>filenameList;
			Result = findRecursiveFile (filenameList, const_cast<char*>(m_arg.c_str()), &r, WS_DEFAULT);
			// get the list of library for all the possible files 
			set<string> librariesList;
			for (auto i : filenameList) Result = getRuntimeLibrariesList(librariesList,i);
			// get the own package  for all the elf files dependencies libraries
#ifndef NDEBUG
			for (auto i : librariesList) cerr << i <<endl;
#endif
			if ( (librariesList.size() > 0 ) && (Result > -1) ) {
				set<string> runtimeList;
				for (set<string>::const_iterator i = librariesList.begin();i != librariesList.end();++i) {
					for (packages_t::const_iterator j = m_listOfInstPackages.begin(); j != m_listOfInstPackages.end();++j) {
						bool found = false;
						for (set<string>::const_iterator k = j->second.files.begin(); k != j->second.files.end(); ++k) {
							if ( k->find('/' + *i) != string::npos) {
								string dependency = j->first + static_cast<ostringstream*>( &(ostringstream() <<  j->second.build ))->str();
								runtimeList.insert(dependency);
								break;
								found = true;
							}
						}
						if ( found == true) {
							found = false;
							break;
						}
					}
				}
				if (runtimeList.size()>0) {
#ifndef NDEBUG
					cerr << "Number of libraries founds: " << runtimeList.size() << endl;
#endif
					unsigned int s = 1;
					for ( auto i : runtimeList ) {
						cout << i << endl;
						s++;
					}
					cout << endl;
				}
			}	
		} else if (m_libraries_mode + m_runtime_mode > 0) {
			// get the list of installed package silently
			buildDatabaseWithDetailsInfos(true);
			set<string> librariesList;
			int Result = -1;
			if (checkPackageNameExist(m_arg)) {
				for (set<string>::const_iterator i = m_listOfInstPackages[m_arg].files.begin();
					i != m_listOfInstPackages[m_arg].files.end();
					++i){
					string filename('/' + *i);
					Result = getRuntimeLibrariesList(librariesList,filename);
				}
				if ( (librariesList.size() > 0 ) && (Result > -1) ) {
					if (m_runtime_mode) {
						set<string> runtimeList;
						for (set<string>::const_iterator i = librariesList.begin();
						i != librariesList.end();
						++i) {
							for (packages_t::const_iterator j = m_listOfInstPackages.begin();
								j != m_listOfInstPackages.end();
								++j){
								bool found = false;
								for (set<string>::const_iterator k = j->second.files.begin();
									k != j->second.files.end();
									++k) {
									if ( k->find('/' + *i) != string::npos) {
										runtimeList.insert(j->first);
										break;
										found = true;
									}
								}
								if (found == true) {
									found = false;
									break;
								}
							}
						}
						if (runtimeList.size()>0) {
							unsigned int s = 1;
							for (set<string>::const_iterator i = runtimeList.begin();
								i!=runtimeList.end();
								++i){
								cout << *i;
								s++;
								if (s <= runtimeList.size())
									cout << ",";
							}
							cout << endl;
						}
					} else {
						for (set<string>::const_iterator i = librariesList.begin();i != librariesList.end();++i)
							cout << *i << endl;
					}
				}
			}	
		} else if (m_epoc) {
			// get the building time of the package return 0 if not found
			buildDatabaseWithDetailsInfos(true);
			if (checkPackageNameExist(m_arg)) {
				cout << m_listOfInstPackages[m_arg].build << endl;
			} else {
				cout << "0" << endl;
			}
		} else if (m_details_mode) {
			// get all the details of a package
			buildDatabaseWithDetailsInfos(false);
			if (checkPackageNameExist(m_arg)) {
				char * c_time_s = ctime(&m_listOfInstPackages[m_arg].build);
				cout << "Name           : " << m_arg << endl
					<< "Description    : " << m_listOfInstPackages[m_arg].description << endl
					<< "URL            : " << m_listOfInstPackages[m_arg].url << endl
					<< "Maintainer(s)  : " << m_listOfInstPackages[m_arg].maintainer << endl
					<< "Packager(s)    : " << m_listOfInstPackages[m_arg].packager << endl
					<< "Version        : " << m_listOfInstPackages[m_arg].version << endl
					<< "Release        : " << m_listOfInstPackages[m_arg].release << endl
					<< "Build date     : " << c_time_s
					<< "Size           : " << m_listOfInstPackages[m_arg].size << endl
					<< "Number of Files: " << m_listOfInstPackages[m_arg].files.size()<< endl
					<< "Arch           : " << m_listOfInstPackages[m_arg].arch << endl;
				if ( m_listOfInstPackages[m_arg].dependencies.size() > 0 ) {
					cout << "Dependencies   : ";
					for ( auto i : m_listOfInstPackages[m_arg].dependencies) cout << i.first << " ";
					cout << endl;
				}
			}
		} else if (m_owner_mode) {
			// List owner(s) of file or directory
			buildDatabaseWithDetailsInfos(false);
			regex_t preg;
			if (regcomp(&preg, m_arg.c_str(), REG_EXTENDED | REG_NOSUB)) {
				m_actualError = CANNOT_COMPILE_REGULAR_EXPRESSION;
				treatErrors(m_arg);
			}
			vector<pair<string, string> > result;
			result.push_back(pair<string, string>("Package", "File"));
			unsigned int width = result.begin()->first.length(); // Width of "Package"
#ifndef NDEBUG
			cerr << m_arg << endl;
#endif
			for (auto i : m_listOfInstPackages) {
				for (auto j : i.second.files) {
					const string file('/' + j);
					if (!regexec(&preg, file.c_str(), 0, 0, 0)) {
						result.push_back(pair<string, string>(i.first, j));
						if (i.first.length() > width) {
							width = i.first.length();
						}
					}
				}
			}

			regfree(&preg);
			if (result.size() > 1) {
				for (auto i : result ) {
					cout << left << setw(width + 2) << i.first << i.second << endl;
				}
			} else {
				cout << m_utilName << ": no owner(s) found" << endl;
			}
		}
	}
}
void Pkginfo::finish()
{
}
void Pkginfo::printHelp() const
{
	cout << USAGE << m_utilName << " [options]" << endl
	     << OPTIONS << endl
	     << "  -i, --installed             list installed packages" << endl
         << "  -d, --details               list details about the <package>" << endl
         << "  -L, --libraries            list all the runtime libraries for the <package>" << endl
	     << "  -l, --list <package|file>   list files in <package> or <file>" << endl
	     << "  -o, --owner <pattern>       list owner(s) of file(s) matching <pattern>" << endl
	     << "  -f, --footprint <file>      print footprint for <file>" << endl
         << "  -a, --archive <file>        print Name, Version, Release, BuildDate and Deps of the <file>" << endl
	     << "  -b, --buildtime <package>   return the name and the build time of the package" << endl
         << "  -R, --runtimedep <package>  return on a single line all the runtime dependencies" << endl
         << "  --runtimedepfiles <path>    return on a single line all the runtime dependencies for the files found in the <path>" << endl
	     << "  -r, --root <path>           specify alternative installation root" << endl
	     << "  -v, --version               print version and exit" << endl
	     << "  -h, --help                  print help and exit" << endl;
}
// vim:set ts=2 :
